feat: new rpc method 'sign' to sign arbitrary data and text messages

PaulBernier opened the pull request:
This PR implements a new capability for the Sollet wallet: a new RPC method allowing to prompt the user to sign text message (string) or arbitrary data (Uint8Array).

## Motivation

Be on a par with MetaMask extension for Ethereum where it is possible to sign data:
![Screenshot from 2021-03-10 20-14-37](https://user-images.githubusercontent.com/5141652/111039359-8c9ed880-83e2-11eb-8014-9e13f7a85d8e.png)
This is useful to prove ownership of an address or sign in with backends (with a better user experience than signing a dummy transaction).

## Changes

- Added handling of a new RPC method `sign` in `PopupPage.js` and introduced `messageType` variable to decide which UI to display.
- Moved most of the existing UI component of `ApproveSignatureForm` into its own component `SignTransactionFormContent` which handles the display for the already existing cases (`signTransaction` and `signAllTransactions`)
- Added a new component `SignTransactionFormContent` to handle the case of signing arbitrary data/text messages.

## Screenshots

Visualization of the two new windows:

![sollet sign message](https://user-images.githubusercontent.com/5141652/111041213-1f904080-83ec-11eb-85fb-7fd8c6f0956b.png)
![sollet sign data](https://user-images.githubusercontent.com/5141652/111041217-21f29a80-83ec-11eb-8685-0f69679f68f5.png)

PaulBernier commented on src/components/SignTransactionFormContent.js:
```diff
@@ -0,0 +1,322 @@
+  return !unsafe;
+}
+
+export default function SignTransactionFormContent({
```
None of this code is new, I just copy-pasted it to extra it into its own component `SignTransactionFormContent`

armaniferrante replied:
Would you mind separating the copy pasted/refactored code into its own commit to make reviewing easier?

PaulBernier replied:
For sure, makes sense. I just split it into two commits, the first one is the refactoring, the second one is the new feature.

armaniferrante reviewed (commented):
LGTM, though there's some merge conflicts to reconcile.

PaulBernier commented:
@armaniferrante to try to minimize future conflicts I made a separate PR with just the refactoring of extracting the component to its own file: https://github.com/project-serum/spl-token-wallet/pull/146 would be nice to have this merged first :)

PaulBernier commented on src/pages/PopupPage.js:
```diff
@@ -104,21 +106,52 @@ export default function PopupPage({ opener }) {
+        messageType = 'tx';
+        break;
+      case 'sign':
+        if (typeof request.params.data === 'string') {
```
The reason I detect the type here is to try to have a better UX when it's a text message, as I am then able in the popup to decide to display as human-readable text. If the input for this method is always  `UInt8Array` then I lose this ability as it's hard to tell if arbitrary data is some utf string or not.
But I think instead of relying directly on the type of `data`, we can have `data` always be a `UInt8Array` and have a second param like `encoding` that hint Sollet how it should be displayed to the user (e.g: `hex`, `utf8`...). I think that would be even better actually! I will start implementing it this way, but let me know if you think otherwise.

PaulBernier replied:
Changed

PaulBernier commented:
@armaniferrante I have rebased this MR on the latest master. I also changed the RPC method to take 2 params: `data`, that must be a `Uint8Array` and `display` that currently can be either `utf8` to display the data as a utf8 encoded string or if not specified will default to `hex`.
